### PR TITLE
[PIR-1187] Pentaho interactive reporting datetime prompt

### PR DESCRIPTION
**Pull request by SergeyTravin**
@mbatchelor, @wseyler could you review?

### CODE REVIEW

**Comment by lucboudreau** [comment]
Changed files:
==============
  src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
  test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java

Checkstyle violations:
----------------------
Errors:File src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
1. Message: object def rcurly at indentation level 8 not at correct indentation, [4, 6] at Line: 250 Column: None
2. Message: object def rcurly at indentation level 8 not at correct indentation, [4, 6] at Line: 256 Column: None
3. Message: object def rcurly at indentation level 8 not at correct indentation, [4, 6] at Line: 262 Column: None
4. Message: object def rcurly at indentation level 8 not at correct indentation, [4, 6] at Line: 268 Column: NoneFile test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java
1. Message: array initialization child at indentation level 8 not at correct indentation, 6 at Line: 42 Column: None

Newly Broken Tests:
=================
Test Case Errors:
-----------------
  
org.pentaho.pms.mql.dialect.DefaultSQLDialectIT.testTimestampType: junit.framework.ComparisonFailure
```
junit.framework.ComparisonFailure: null expected:<'2015-10-22 [18]:54:58.252'> but was:<'2015-10-22 [06]:54:58.252'>
	at org.pentaho.pms.MetadataTestBase.assertEqualsIgnoreWhitespaces(MetadataTestBase.java:61)
	at org.pentaho.pms.mql.dialect.DefaultSQLDialectIT.testTimestampType(DefaultSQLDialectIT.java:50)

```

Still Broken Tests:
===================
Test Case Errors:
-----------------
  org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization(SQLModelGeneratorIT.java:161)

```org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion: org.pentaho.metadata.util.SQLModelGeneratorException
```
org.pentaho.metadata.util.SQLModelGeneratorException: SQLModelGenerator.ERROR_0001 - Input Validation Failed
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:237)
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:125)
	at org.pentaho.metadata.SQLModelGeneratorIT.generateModel(SQLModelGeneratorIT.java:274)
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion(SQLModelGeneratorIT.java:168)

```org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy(SQLModelGeneratorIT.java:97)

```org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator(SQLModelGeneratorIT.java:81)

```

Integration test coverage change:
============
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange :**-12.50%**
 * complexityPercentCoverageChange :**-12.50%**
    
org.pentaho.metadata.util.SQLModelGenerator:
 * instructionPercentCoverageChange : +0.36%
 * branchPercentCoverageChange : +2.70%
 * linePercentCoverageChange : +0.75%
 * complexityPercentCoverageChange : +2.04%
    
org.pentaho.pms.mql.dialect.DefaultSQLDialect:
 * linePercentCoverageChange : +0.03%

**Comment by SergeyTravin** [comment]
@lucboudreau 
I have fixed tests and checkstyle violations, could you review it again?

Great Thanks!

**Comment by brosander** [comment]
Changed files:
==============
  src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
  test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java

Checkstyle violations:
----------------------
No errors found

Newly Broken Tests:
=================
Test Case Errors:
-----------------
  
org.pentaho.pms.mql.dialect.DefaultSQLDialectIT.testTimestampType: junit.framework.ComparisonFailure
```
junit.framework.ComparisonFailure: null expected:<'1970-01-01 [12]:00:00.0'> but was:<'1970-01-01 [00]:00:00.0'>
	at org.pentaho.pms.MetadataTestBase.assertEqualsIgnoreWhitespaces(MetadataTestBase.java:61)
	at org.pentaho.pms.mql.dialect.DefaultSQLDialectIT.testTimestampType(DefaultSQLDialectIT.java:56)

```

Still Broken Tests:
===================
Test Case Errors:
-----------------
  org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization(SQLModelGeneratorIT.java:161)

```org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion: org.pentaho.metadata.util.SQLModelGeneratorException
```
org.pentaho.metadata.util.SQLModelGeneratorException: SQLModelGenerator.ERROR_0001 - Input Validation Failed
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:237)
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:125)
	at org.pentaho.metadata.SQLModelGeneratorIT.generateModel(SQLModelGeneratorIT.java:274)
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion(SQLModelGeneratorIT.java:168)

```org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy(SQLModelGeneratorIT.java:97)

```org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator(SQLModelGeneratorIT.java:81)

```
Unit test coverage change:
============
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange :**-12.50%**
 * complexityPercentCoverageChange :**-12.50%**
    
Integration test coverage change:
============
org.pentaho.metadata.util.SQLModelGenerator:
 * instructionPercentCoverageChange : +0.36%
 * branchPercentCoverageChange : +2.70%
 * linePercentCoverageChange : +0.75%
 * complexityPercentCoverageChange : +2.04%
    
org.pentaho.pms.mql.dialect.DefaultSQLDialect:
 * linePercentCoverageChange : +0.03%

**Comment by SergeyTravin** [comment]
@brosander  Please review when you get a chance.

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Finally:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
  test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java

Checkstyle violations:
======================
No errors found

Newly Broken Tests:
-----------------
org.pentaho.pms.mql.dialect.DefaultSQLDialectIT.testTimestampType: junit.framework.ComparisonFailure
```
junit.framework.ComparisonFailure: null expected:<'1970-01-01 [12]:00:00.0'> but was:<'1970-01-01 [00]:00:00.0'>
	at org.pentaho.pms.MetadataTestBase.assertEqualsIgnoreWhitespaces(MetadataTestBase.java:61)
	at org.pentaho.pms.mql.dialect.DefaultSQLDialectIT.testTimestampType(DefaultSQLDialectIT.java:56)

```



Still Broken Tests:
-----------------
org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization(SQLModelGeneratorIT.java:161)

```
org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion: org.pentaho.metadata.util.SQLModelGeneratorException
```
org.pentaho.metadata.util.SQLModelGeneratorException: SQLModelGenerator.ERROR_0001 - Input Validation Failed
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:237)
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:125)
	at org.pentaho.metadata.SQLModelGeneratorIT.generateModel(SQLModelGeneratorIT.java:274)
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion(SQLModelGeneratorIT.java:168)

```
org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy(SQLModelGeneratorIT.java:97)

```
org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator(SQLModelGeneratorIT.java:81)

```

Unit test coverage change:
============
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange : +12.50%
 * complexityPercentCoverageChange : +12.50%
    
Integration test coverage change:
============
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange : +12.50%
 * complexityPercentCoverageChange : +12.50%
    
org.pentaho.pms.mql.dialect.DefaultSQLDialect:
 * linePercentCoverageChange : +0.03%

**Comment by lucboudreau** [comment]
Some tests are still failing. Please see the report above.

**Comment by SergeyTravin** [comment]
@lucboudreau I have  resolved the broken tests. Could you review when you get a chance. Thanks!

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Finally:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  ivy.xml
  src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
  test-src/org/pentaho/helpers/SQLDialectHelper.java
  test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java

Checkstyle violations:
======================
No errors found





Still Broken Tests:
-----------------
org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization(SQLModelGeneratorIT.java:161)

```
org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion: org.pentaho.metadata.util.SQLModelGeneratorException
```
org.pentaho.metadata.util.SQLModelGeneratorException: SQLModelGenerator.ERROR_0001 - Input Validation Failed
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:237)
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:125)
	at org.pentaho.metadata.SQLModelGeneratorIT.generateModel(SQLModelGeneratorIT.java:274)
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion(SQLModelGeneratorIT.java:168)

```
org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy(SQLModelGeneratorIT.java:97)

```
org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator(SQLModelGeneratorIT.java:81)

```


Integration test coverage change:
============
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange :**-12.50%**
 * complexityPercentCoverageChange :**-12.50%**
    
org.pentaho.pms.mql.dialect.DefaultSQLDialect:
 * branchPercentCoverageChange : +0.58%
 * linePercentCoverageChange : +0.03%
 * complexityPercentCoverageChange : +0.81%

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Finally:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  ivy.xml
  src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
  test-src/org/pentaho/helpers/SQLDialectHelper.java
  test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java

Checkstyle violations:
======================
No errors found





Still Broken Tests:
-----------------
org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization(SQLModelGeneratorIT.java:161)

```
org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion: org.pentaho.metadata.util.SQLModelGeneratorException
```
org.pentaho.metadata.util.SQLModelGeneratorException: SQLModelGenerator.ERROR_0001 - Input Validation Failed
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:237)
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:125)
	at org.pentaho.metadata.SQLModelGeneratorIT.generateModel(SQLModelGeneratorIT.java:274)
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion(SQLModelGeneratorIT.java:168)

```
org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy(SQLModelGeneratorIT.java:97)

```
org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator(SQLModelGeneratorIT.java:81)

```


Integration test coverage change:
============
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange :**-12.50%**
 * complexityPercentCoverageChange :**-12.50%**
    
org.pentaho.pms.mql.dialect.DefaultSQLDialect:
 * branchPercentCoverageChange : +0.58%
 * linePercentCoverageChange : +0.03%
 * complexityPercentCoverageChange : +0.81%

**Comment by SergeyTravin** [comment]
@lucboudreau The PR does not have any checkstyle violations or failed tests. It increases a coverage for changed file. I invistigated  org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList and does not found any relations to my PR. Could you give any comments?

**Comment by lucboudreau** [comment]
There are still 4 test failures. Please see if you can fix them.

**Comment by SergeyTravin** [comment]
@lucboudreau, There are still 4 test failures. Please see if you can fix them.

Tests were fixed. Could you check it if you have a chance, please.

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. ant -f build.xml clean-all resolve jacoco jacoco-integration checkstyle

Cleanup:
-------
rm -r ~/.ivy2/local || echo "no publish local to remove"

Changed files:
==============
  ivy.xml
  src/org/pentaho/pms/mql/dialect/DefaultSQLDialect.java
  test-res/org/pentaho/metadata/sampledata.properties
  test-res/org/pentaho/metadata/sampledata.script
  test-src/org/pentaho/helpers/SQLDialectHelper.java
  test-src/org/pentaho/metadata/SQLModelGeneratorIT.java
  test-src/org/pentaho/metadata/model/olap/util/OlapUtilIT.java
  test-src/org/pentaho/pms/mql/dialect/DefaultSQLDialectIT.java
  test-src/org/pentaho/pms/mql/dialect/HiveDialectIT.java

Checkstyle violations:
======================
No errors found

License header violations:
==========================
No violations found




Newly Fixed Tests:
-----------------
org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryXmlSerialization(SQLModelGeneratorIT.java:161)

```
org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion: org.pentaho.metadata.util.SQLModelGeneratorException
```
org.pentaho.metadata.util.SQLModelGeneratorException: SQLModelGenerator.ERROR_0001 - Input Validation Failed
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:237)
	at org.pentaho.metadata.util.SQLModelGenerator.generate(SQLModelGenerator.java:125)
	at org.pentaho.metadata.SQLModelGeneratorIT.generateModel(SQLModelGeneratorIT.java:274)
	at org.pentaho.metadata.SQLModelGeneratorIT.testQueryConversion(SQLModelGeneratorIT.java:168)

```
org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testToLegacy(SQLModelGeneratorIT.java:97)

```
org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
	at org.pentaho.metadata.SQLModelGeneratorIT.testSQLModelGenerator(SQLModelGeneratorIT.java:81)

```




Integration test coverage change:
============
org.pentaho.metadata.util.SQLModelGenerator:
 * instructionPercentCoverageChange : +66.24%
 * branchPercentCoverageChange : +32.43%
 * linePercentCoverageChange : +52.99%
 * complexityPercentCoverageChange : +12.24%
 * methodPercentCoverageChange : +14.29%
    
org.pentaho.metadata.query.model.util.CsvDataTypeEvaluator.CountItemsList:
 * branchPercentCoverageChange : +12.50%
 * complexityPercentCoverageChange : +12.50%
    
org.pentaho.metadata.util.ThinModelConverter:
 * instructionPercentCoverageChange : +4.22%
 * branchPercentCoverageChange : +5.50%
 * linePercentCoverageChange : +4.26%
 * complexityPercentCoverageChange : +4.42%
    
org.pentaho.pms.mql.dialect.DefaultSQLDialect:
 * branchPercentCoverageChange : +0.58%
 * linePercentCoverageChange : +0.03%
 * complexityPercentCoverageChange : +0.81%
    
org.pentaho.pms.locale.Locales:
 * branchPercentCoverage :25.00%
 * instructionPercentCoverageChange : +14.57%
 * linePercentCoverageChange : +9.76%
 * complexityPercentCoverageChange : +10.00%
 * methodPercentCoverageChange : +6.25%
    
org.pentaho.pms.util.Settings:
 * instructionPercentCoverageChange : +5.15%
 * linePercentCoverageChange : +3.45%
 * complexityPercentCoverageChange : +5.88%
 * methodPercentCoverageChange : +5.88%

File: ivy.xml [hunk 1]
```diff
@@ -66,7 +66,7 @@
     <dependency org="${dependency.libbase.group}" name="libbase" rev="${dependency.libbase.revision}" />
     
     <!-- test dependencies -->
-    <dependency org="junit" name="junit" rev="4.4" conf="test->default" />
+    <dependency org="junit" name="junit" rev="4.12" conf="test->default" />
```
Comment (lucboudreau) [inline ivy.xml]: This is verified. Current version to use in platform is 4.12.

File: test-res/org/pentaho/metadata/sampledata.properties [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#HSQL Database Engine 2.3.2
```
Comment (lucboudreau) [inline test-res/org/pentaho/metadata/sampledata.properties]: Was this a mistake? Not sure we're supposed to update this. Please verify.

Comment (SergeyTravin) [inline test-res/org/pentaho/metadata/sampledata.properties]: It is not mistake.  Please see
https://github.com/pentaho/pentaho-metadata/blob/master/ivy.xml#L79

You are completely right, that the changes can be ignored.  The file was added to .gitignore.

**Pull request merged**